File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    
    //create set
    //loop thru string
    //conditional if set has letter
    //if doesnt then add to set
    //find how big the set is
    
//     let set = new Set();
//     let counter = 0
//     let maxSize = 0
    
//     for(let i = 0; i < s.length; i++){
//         if(set.has(s[i])){
//             set.delete(s[counter])
//             counter++
//         }
//         set.add(s[i])
//         maxSize = Math.max(maxSize, set.size)
        
//     }
//     return maxSize
    
    const set = new Set();
    let l = 0;
    let max = 0;

    for (let r = 0; r < s.length; r++) {
        while (set.has(s[r])) {
            set.delete(s[l]);
            l++;
        }
        set.add(s[r]);
        max = Math.max(max, set.size);
    }
    return max;
    
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = set()
        count = 0
        [r, l] = [0,0]
        
        if len(s) == 1:
            return 1
        elif len(s) == 0:
            return 0
        
        
        while r < len(s):
            if s[r] not in res:
                res.add(s[r])
            else:
                count = max(count, len(res))
                while s[r] in res:
                    res.remove(s[l])
                    l+=1
                res.add(s[r])
            r+=1
        
        count = max(count, len(res))

        
        return count
                